### DISCUSSION TITLE
data from useMutation hook is always undefined

### DISCUSSION THREAD

**Discussion by fryck** [urql-graphql/urql#2900] [Q&A]
Hello, 

Hopefully you could help me as I stuck on that since several days...

I have an issue using hook `useMutation`, the data is always undefined, when I use graphcache, here is my graphql schema (partial):

```
type Quote {
  advisorId: String
  createdAt: DateTime!
  customDeliveryPrice: Float
  customPrice: Float
  customPriceTextVariable: Float
  customer: Customer!
  customerComment: String
  deliveryLocation: DeliveryLocation!
  deliveryPrice: Float!
  id: ID!
  orderForm: OrderForm
  parentId: String
  paymentMode: [PaymentMode!]!
  price: Float!
  priceTextVariable: Float!
  products: [QuoteProduct!]!
  salesCsltComment: String
  status: QuoteStatus
  statusIcon: [IconStatus!]!
  updatedAt: DateTime
  wishedDeliveredDate: DateTime
}

type Customer {
  corporateName: String
  createdAt: DateTime!
  email: String!
  firstName: String!
  id: ID!
  lastName: String!
  phoneNumber: String
  updatedAt: DateTime
}

type QuoteProduct {
  category: String!
  colors: [Json!]!
  createdAt: DateTime!
  customPrice: Float
  customTotalPrice: Float
  customTotalPriceTextVariable: Float
  customerComment: String
  customizations: [QuoteProductCustomization!]!
  deletedAt: DateTime
  id: ID!
  image: String!
  markingTechnique: MarkingTechnique!
  name: String!
  outfits: [QuoteProductOutfit!]!
  price: Float!
  projectTitle: String!
  quantity: Int!
  quote: Quote!
  salesCsltComment: String
  shippingDelay: String!
  sku: String!
  totalPrice: Float!
  totalPriceTextVariable: Float
  updatedAt: DateTime
  validityDate: DateTime!
  wc_product: WCProduct!
}

type QuoteProductCustomization {
  colorCount: String
  createdAt: DateTime!
  customPriceLogo: Float
  customPriceText: Float
  customPriceTextVariable: Float
  customerComment: String
  deletedAt: DateTime
  id: ID!
  logo: QuoteProductCustomizationLogo
  position: MarkingPosition!
  priceLogo: Float
  priceText: Float
  priceTextVariable: Float
  quoteProduct: QuoteProduct!
  salesCsltComment: String
  text: String
  textVariable: Boolean
  updatedAt: DateTime
}

type QuotePaginateType {
  _total: Int
  data: [Quote!]
}

type Query {
  quote(id: ID!): Quote
  quotes(orderBy: [QuoteOrderByInput!] = {createdAt: desc}, skip: Int, take: Int, where: QuoteWhereInput): QuotePaginateType
}

type Mutation {
  updateQuote(id: ID!, input: QuoteUpdateInputType): Quote
}
```

In my list component, I query for all quotes (I'm using urql graphql codegen plugin, to generate hooks):

```
const Quotes = () => {
  const [{ fetching, data }] = useQuotesQuery({
    variables: {
      take: 50,
      skip: 0,
    },
  });
}
export default withUrqlClient(
  (_ssrExchange) => ({
    url: GRAPHQL_API_ENDPOINT,
    exchanges: [
      dedupExchange,
      cacheExchange({}),
      _ssrExchange,
      multipartFetchExchange,
    ],
  }),
  { ssr: false },
)(Quotes);
```

And when I trigger update mutation on a particular quote entity :

```
mutation updateQuote($updateQuoteId: ID!, $input: QuoteUpdateInputType) {
  updateQuote(id: $updateQuoteId, input: $input) {
    id
    createdAt
    updatedAt
    deliveryLocation
    wishedDeliveredDate
    customerComment
    salesCsltComment
    statusIcon
    status
    price
    customPrice
    priceTextVariable
    customPriceTextVariable
    deliveryPrice
    customDeliveryPrice
    paymentMode
    parentId
    advisorId
    customer {
      id
      email
      firstName
      lastName
      phoneNumber
      corporateName
      createdAt
      updatedAt
    }
    products {
      id
      sku
      projectTitle
      category
      name
      price
      customPrice
      totalPrice
      customTotalPrice
      totalPriceTextVariable
      customTotalPriceTextVariable
      shippingDelay
      salesCsltComment
      outfits {
        path
        url
      }
      wc_product {
        id
      }
      colors
      customerComment
      customizations {
        id
        position
        text
        textVariable
        logo {
          path
          url
        }
        priceLogo
        customPriceLogo
        priceText
        customPriceText
        priceTextVariable
        customPriceTextVariable
        colorCount
        customerComment
        salesCsltComment
      }
      quantity
      markingTechnique
      image
      validityDate
    }
  }
}

const [
    {
      data: updateQuoteData,
      error: updateQuoteError,
      fetching: updatingQuote,
    },
    updateQuote,
  ] = useUpdateQuoteMutation();
  
  
 console.log(updateQuoteData); // always undefined
 console.log(updatingQuote); // its working
 console.log(updateQuoteError); // No Error
```

Mutation response:

```
{
    "data": {
        "updateQuote": {
            "id": "2599237a-db47-479d-86b1-37c2317f8fad",
            "createdAt": "2022-12-19T16:46:41.420Z",
            "updatedAt": "2022-12-21T11:07:40.731Z",
            "deliveryLocation": "FRANCE",
            "wishedDeliveredDate": "2022-12-19T00:00:00.000Z",
            "customerComment": "Test",
            "salesCsltComment": "",
            "statusIcon": [],
            "status": null,
            "price": 755,
            "customPrice": null,
            "priceTextVariable": 755,
            "customPriceTextVariable": null,
            "deliveryPrice": 0,
            "customDeliveryPrice": null,
            "paymentMode": [
                "BANK_TRANSFER"
            ],
            "parentId": null,
            "advisorId": null,
            "customer": {
                "id": "026ef3c5-08b4-466b-a5d9-aefd0063050d",
                "email": "test@test.fr",
                "firstName": "Fabien",
                "lastName": "Ryckoort",
                "phoneNumber": "0671509203",
                "corporateName": "Test",
                "createdAt": "2022-12-19T16:46:41.420Z",
                "updatedAt": "2022-12-21T11:07:40.728Z",
                "__typename": "Customer"
            },
            "products": [
                {
                    "id": "70081efd-40e5-44e5-8155-4db7e3d9789e",
                    "sku": "FY05016",
                    "projectTitle": "",
                    "category": "Casquettes personnalisées",
                    "name": "Anderlecht",
                    "price": 6.5,
                    "customPrice": null,
                    "totalPrice": 6.5,
                    "customTotalPrice": null,
                    "totalPriceTextVariable": 6.5,
                    "customTotalPriceTextVariable": null,
                    "shippingDelay": "3 à 4 semaines",
                    "salesCsltComment": "",
                    "outfits": [],
                    "wc_product": {
                        "id": 21062,
                        "__typename": "WCProduct"
                    },
                    "colors": [
                        {
                            "name": "Black",
                            "image": "https://flockyou.fr/wp-content/uploads/anderlecht/Beechfield-B646-Black-Product-1000.jpg"
                        }
                    ],
                    "customerComment": "",
                    "customizations": [],
                    "quantity": 100,
                    "markingTechnique": "EMBROIDERY",
                    "image": "https://flockyou.fr/wp-content/uploads/anderlecht/Beechfield-B646-Olive-Green-Product-1000.jpg",
                    "validityDate": "2023-02-17T00:00:00.000Z",
                    "__typename": "QuoteProduct"
                },
                {
                    "id": "fecc4232-6637-496b-8fb0-9968a5bbc039",
                    "sku": "FY06017",
                    "projectTitle": "",
                    "category": "Bonnets personnalisés",
                    "name": "Courchevel",
                    "price": 10.5,
                    "customPrice": null,
                    "totalPrice": 10.5,
                    "customTotalPrice": null,
                    "totalPriceTextVariable": 10.5,
                    "customTotalPriceTextVariable": null,
                    "shippingDelay": "",
                    "salesCsltComment": "",
                    "outfits": [],
                    "wc_product": {
                        "id": 23700,
                        "__typename": "WCProduct"
                    },
                    "colors": [
                        {
                            "name": "Black",
                            "image": "https://flockyou.fr/wp-content/uploads/courchevel/bc432_black_ft.jpg"
                        }
                    ],
                    "customerComment": "",
                    "customizations": [],
                    "quantity": 10,
                    "markingTechnique": "NONE",
                    "image": "https://flockyou.fr/wp-content/uploads/courchevel/bc432_pastelpink_ft.jpg",
                    "validityDate": "2023-02-18T00:00:00.000Z",
                    "__typename": "QuoteProduct"
                }
            ],
            "__typename": "Quote"
        }
    }
}
```

As you can see, all `__typename` is correctly assigned, and all entity have an ID

When my mutations is triggered, my entity `Quote(id)` is correctly updated in cache (and its relations) as I can see my data is updated in the quotes listing component and even in the update form (quote query).

I noticed that when I remove all `id` field in `updateQuote` mutation, the data from hook is populated, otherwise no.

I'm unable to reproduce the issue with same dependencies version etc. in a simple Todo app example : https://codesandbox.io/s/affectionate-bartik-bfnm0g?file=/pages/index.js

Could you give me an idea on what can cause this issue with graphcache ?

Many Thanks !

**Comment by JoviDeCroock** [comment]
Do you have a warning in the console about certain entities not being keyable, as you have non-keyable entities without a `keys` config? Or have you by any chance omitted the updaters where you had a `cache.invalidate()` call?

**Reply by fryck** [reply]
Thanks for your reply @JoviDeCroock, so I checked you suggestion and I have no warnings about non-cacheable entities, here is my keys definition for graphcache:

```
keys: {
          Quote: (data) => {
            return String(data.id);
          },
          CustomizationPosition: (data) => {
            return String(data.position);
          },
          QuoteProductOutfit: (data) => {
            return `${String(data.path)}${
              data.versionId ? `-${data.versionId}` : ''
            }`;
          },
          QuoteProductCustomizationLogo: (data) => {
            return `${String(data.path)}${
              data.versionId ? `-${data.versionId}` : ''
            }`;
          },
          QuotePaginateType: ({ take, skip }) => {
            return `${skip}-${take}`;
          },
        },
 ```
 
 To be sure that, it's not an issue with relation entity of Quote à remove a lot of fields to test, so for my query `quote($id)` I return only this fields:
 
 ```
 query quote($quoteId: ID!) {
  quote(id: $quoteId) {
    id
    deliveryPrice
    customDeliveryPrice
  }
}
```

and for the mutation response (update Quote entity) I return the same fields:

```
mutation updateQuote($updateQuoteId: ID!, $input: QuoteUpdateInputType) {
  updateQuote(id: $updateQuoteId, input: $input) {
    id
    deliveryPrice
    customDeliveryPrice
  }
}
```

I have the same issue, I really dont understand what can cause this issue, but I'm pretty sure it's related to graphcache.

I don't use cache updaters for my mutations, so I don't have `cache.invalidate` call.

Like I said, my `Quote` entity is correctly updated to cache after mutation.

Thanks for your help !

**Reply by JoviDeCroock** [reply]
I wonder if it's because most of those keys can't be achieved with the payload you are using with your mutation, it's okay to embed entities like:

A key is supposed to make an entity unique, if it can't be unique or you stop querying the key it's better to not register a key and embed the entity.

```js
keys: {
        Quote: (data) => {
          // Using id happens automatically, can remove this
          return String(data.id);
        },
        CustomizationPosition: (data) => {
          // can probably just return null here because multiple parents will have i.e.
          // position 0 on an entity
          return null
        },
        QuoteProductOutfit: (data) => {
          // if this doesn't make it unique/you are not always querying this you can remove
          // this duplicates entities based on your selection-set which is dangerous
          return null;
        },
        QuoteProductCustomizationLogo: (data) => {
          // same as above
          return null;
        },
        QuotePaginateType: ({ take, skip }) => {
          // You will def have pages who use the same skip and take across entities, so also dangerous
          return null;
        },
},
```

**Reply by fryck** [reply]
Thanks, I applied your recommandation for keys.

I just try to call `updateQuote` mutation on a Quote BEFORE querying it, so the entity is not in cache, and it works...

> I wonder if it's because most of those keys can't be achieved with the payload you are using with your mutation

Yes it's probably the cause, but I don't know how to find the key responsible to that issue.

I've tried to remove a lot of data from `query Quote($id)` payload and `mutation updateQuote()` payload, but no success

**Reply by fryck** [reply]
I've compared the payload of query response and mutation response, it's exactly the same, all child entities have an `id` and `__typename`

**Reply by JoviDeCroock** [reply]
You can probably use some `debugExchange` before and after the `cacheExchange` to see how stream traffic looks but this really feels like something subtile that you have going on for yourself, especially if you can't reproduce it 😅

**Reply by fryck** [reply]
I think I'm not doing something "special" tbh ^^

**Reply by fryck** [reply]
I've checked with `debugExhange` and I can see the data in the log for my mutation

![Screenshot 2022-12-23 at 10 57 09](https://user-images.githubusercontent.com/1364225/209314861-87698b2c-2fc2-49db-960f-1a6c80f21bf1.png)

**Reply by JoviDeCroock** [reply]
I mean... where did you put that `debugExchange` 😅 I mean, anyway, if this is a bug we need a reproduction. Happy to try and help but I can't predict what you're doing

**Reply by fryck** [reply]
I put it in the `exchanges` array :

```
export default withUrqlClient(
  (_ssrExchange) => ({
    url: GRAPHQL_API_ENDPOINT,
    exchanges: [
      debugExchange,
      dedupExchange,
      cacheExchange({...}),
      _ssrExchange,
      multipartFetchExchange,
    ],
  }),
  { ssr: false },
)(Quotes);
```
 
I'm able to reproduce the issue with a simple delete mutation:

```
type QuoteProductDeleteResponse {
  deletedAt: DateTime
  id: ID!
  quoteId: String!
}

type Mutation {
  deleteQuoteProduct(id: ID!): QuoteProductDeleteResponse
}

mutation deleteQuoteProduct($id: ID!) {
  deleteQuoteProduct(id: $id) {
    id
    deletedAt
    quoteId
  }
}
```
I really don't understand, and I don't know how you can reproduce it on your side...

**Reply by fryck** [reply]
I think I found issue

When I query quotes list

```
type QuotePaginate {
  _total: Int
  items: [Quote!]
}

type Query {
  quotes(orderBy: [QuoteOrderByInput!] = {createdAt: desc}, skip: Int, take: Int, where: QuoteWhereInput): QuotePaginate
}

query quotes($take: Int, $skip: Int, $orderBy: [QuoteOrderByInput!]) {
  quotes(take: $take, skip: $skip, orderBy: $orderBy) {
    _total
    skip
    take
    items {
      id
      customer {
         id
         email
         firstName
         lastName
         corporateName
         phoneNumber
       }
      createdAt
      updatedAt
      wishedDeliveredDate
      products {
         id
         quantity
      }
      statusIcon
    }
  }
}
```

If I create a dedicated `QuotePageItem` type, instead of using `[Quote!]` in the `QuotePaginate` type, it works... but I not understand why.

I tried to reproduce the same scenario here: https://codesandbox.io/s/affectionate-bartik-bfnm0g?file=/pages/index.js

**Reply by JoviDeCroock** [reply]
Check the warnings in the console 😅 does your app have those as well? Because if you stop querying a way to achieve the key like you do with `todos.author`, ... then unexpected behavior is to be expected

**Reply by fryck** [reply]
I have 0 warnings in console.

> Because if you stop querying a way to achieve the key like you do with todos.author

Could you give me more details about that ? I'm not sure to understand

**Reply by JoviDeCroock** [reply]
All documented [here](https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/#normalizing-relational-data)

TLDR: we look for `id` and `_id`, if you don't have those as part of your selection-set we can't make a key for your entity and not normalize it. If it can't be keyed you will be warned and it's your responsibility to either return a key or return null in the `keys` config. If you provide a function but don't return anything from it because i.e. you didn't put it in your selection set we can't do much, example your `String(data.id)` becomes "undefined" because you didn't query it (generally better to return `data.id` so we can warn you about undefined)

**Reply by fryck** [reply]
Yes, I already read the documentation about this topic (a lot :D)

As you can see, all my entities have an `id` and `__typename`, for example:

```
{
    "data": {
        "quote": {
            "id": "447ccf33-cd32-4314-ba57-037be0cdfdf1",
            "createdAt": "2022-12-23T15:29:47.983Z",
            "updatedAt": "2022-12-23T17:21:19.760Z",
            "deliveryLocation": "FRANCE",
            "wishedDeliveredDate": "2022-12-23T00:00:00.000Z",
            "customerComment": "",
            "salesCsltComment": "",
            "status": null,
            "price": 0,
            "customPrice": null,
            "priceTextVariable": 0,
            "customPriceTextVariable": null,
            "deliveryPrice": 0,
            "customDeliveryPrice": null,
            "paymentMode": [
                "ONLINE",
                "CHECK"
            ],
            "parentId": null,
            "advisorId": null,
            "customer": {
                "id": "37534db7-4681-42e0-a844-3ffc6cf077d1",
                "email": "test@test.com",
                "firstName": "SFsdfsd",
                "lastName": "sdfsdg",
                "phoneNumber": "22848234",
                "corporateName": "Aerhw",
                "createdAt": "2022-12-23T15:29:47.983Z",
                "updatedAt": "2022-12-23T17:21:19.756Z",
                "__typename": "Customer"
            },
            "products": [],
            "__typename": "Quote"
        }
    }
}
```

I always return an id
